Developers: add permissionless contract documentation (#10)
alexgo-io/alex-v1-docs: developers/protocol-contracts/self-listing-helper-v3a.clar.md

# self-listing-helper-v3a.clar

- [Deployed contract](https://explorer.hiro.so/txid/SP1E0XBN9T4B10E9QMR7XMFJPMA19D77WY3KP2QKC.self-listing-helper-v3a?chain=mainnet)

The `self-listing-helper-v3a` contract enables the creation of trading pools on the ALEX DEX through two distinct mechanisms: a permissioned flow and a permissionless flow.

> In the **permissioned flow**, pre-approved tokens can create pools by following a guided process through the ALEX UI. These tokens must be whitelisted and have sufficient anchor token liquidity. Each pool must pair a pre-approved anchor token (`token-x`) with a user-provided token (`token-y`), which is the asset being newly listed.

Reviewer: ```suggestion
In the **permissioned flow**, pools can be created with pre-approved tokens by following a guided process through the ALEX UI. Each pool must pair an **anchor token** (`token-x`) with a user-provided token (`token-y`), which is the asset being newly listed (**listed token**). Both anchor and listed tokens must be approved by the ALEX team.
```

> In the **permissionless flow**, any user can list a new token (`token-y`) by deploying a wrapper contract that matches an approved template. The pool is formed against an existing, governance-approved `token-x`. The contract includes a verification system to ensure the integrity of the deployment: it reconstructs and validates the original contract transaction using Merkle proofs and block data, confirming that the wrapper is trustworthy.

Reviewer: ```suggestion
In the **permissionless flow**, users can list a new token (`token-y`) by deploying a wrapper contract that matches an approved template, and without requiring the approval step. The pool is formed against an existing, governance-approved `token-x`. The contract includes a verification system to ensure the integrity of the deployment: it reconstructs and validates the original contract transaction using Merkle proofs and block data, confirming that the wrapper is trustworthy.
```

Pregunta sobre la segunda oración de ese párrafo: los `token-x` habilitados (anchor tokens) son los mismos para ambos flujos? En ese caso, reorganizaría algunas definiciones.

Author: si, los mismos

> Once verification succeeds, the token is dynamically approved and a pool is created.

Reviewer: Esta oración es parte de permisionless?

Author: si si, me quedo separada se ve

The contract also supports:

- Liquidity locking or burning for pool integrity
- On-chain parameter configuration (fees, oracles, thresholds)
- Governance-controlled token approvals
- Rebates management for AMM incentives

This dual approach gives projects the flexibility to choose between a guided listing process or a fully autonomous, on-chain setup — all while using the same core infrastructure.

## Features

> ### Public Features

Reviewer: ```suggestion
### Public
```

#### `create`

> This function initiates the creation of a liquidity pool between two pre-approved tokens. It is used in the permissioned listing flow, where the listing token (`token-y`) has already been approved by governance or whitelisted in the system.

Reviewer: Si arriba se introduce el concepto de anchor y listing token, esa oración puede cambiarse a algo como:

```suggestion
Permissioned pool creation. Listed token must be pre-approved to create the new pool.
```

Vos ponés
```
This function initiates the creation of a liquidity pool between two pre-approved tokens.
```
pero me parece imporante que se entienda que el anchor siempre es pre-aprobado. Más aún, el set de anchor tokens es elegido por governanza. El listed (o _quoted_ le dicen a veces) token es también supongo que aprobado por gobernanza pero es distinto, por que se aprueba por petición de un usuario. Los anchors tokens tienen que ser tokens de confianza. Bueno nada, lo que ya sabés de ALEX Lab. Me interesa que no estén al mismo nivel tipo "two pre-approved tokens".

The function first runs validation checks via `pre-check`, which ensures the anchor token (`token-x`) is approved and that the caller provides sufficient liquidity. It also checks that no existing pool with the given token pair and `factor` already exists.

Next, it verifies that the listing token has a reserve in the `amm-vault-v2-01` contract, which serves as a proxy for its approval status.

> If all validations pass, the function calls `post-check`, which:

Reviewer: Poner link a `post-check` dentro del file. Evaluar si hace falta repetir data que ya está en `post-check`? 

Este comentario aplica para toda función interna que se mencione!

- Configures its parameters (fees, thresholds, oracle)
- Applies LP token lock or burn rules via `liquidity-locker` if requested

The function emits a print log with the full pool configuration for transparency.

##### Parameters

| Name              | Type                                                                                                                                                                                                                                                                                                 |
| ----------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `request-details` | `{ token-x-trait: <ft-trait>, token-y-trait: <ft-trait>, factor: uint, bal-x: uint, bal-y: uint, fee-rate-x: uint, fee-rate-y: uint, max-in-ratio: uint, max-out-ratio: uint, threshold-x: uint, threshold-y: uint, oracle-enabled: bool, oracle-average: uint, start-block: uint, lock: (buff 1) }` |

#### `create2`

> This function allows the creation of a new liquidity pool using a token that has not been pre-approved by governance. It is the main entry point for the permissionless listing flow.

Reviewer: Same que el comentario en `create`. Sería correcto poner algo así? El pool se crea automáticamente?

```suggestion
Permissionless pool creation. The pool is created in one-transaction step by using .... 
```

- Transfers the LP tokens back to the user
- Deletes the corresponding lock record

##### Parameters

| Name      | Type   |
| --------- | ------ |
| `pool-id` | `uint` |

### Governance Features

#### `is-dao-or-extension`

This standard protocol function checks whether a caller (`tx-sender`) is the DAO executor or an authorized extension, delegating the extensions check to the `executor-dao` contract.

#### `approve-token-x`

A public function, governed through the `is-dao-or-extension`, that sets the approval status and minimum required balance for a token to be used as the anchor token (`token-x`) in pool creation.

This function updates the `approved-token-x` map, enabling the protocol to define which tokens can serve as anchors and to enforce a minimum contribution threshold (`min-x`) for those tokens.

##### Parameters

| Name       | Type        |
| ---------- | ----------- |
| `token`    | `principal` |
| `approved` | `bool`      |
| `min-x`    | `uint`      |

#### `set-fee-rebate`

A public function, governed through the `is-dao-or-extension`, that sets the default fee rebate value used during pool creation.

> This value is stored locally in the contract and passed as an argument to the `set-fee-rebate` function of the external `amm-registry-v2-01` contract when a new pool is initialized.

Reviewer: Sacaría el **external** de los contratos que mencionamos. Me parece que hace alusión a que es un contrato externo al protocolo, cuando son parte del mismo sistema. A lo mejor en liquidity-locker tiene más sentido llamarlo external... aunque entiendo que es codeado por ellos.

# self-listing-helper-v3a.clar

- [Deployed contract](https://explorer.hiro.so/txid/SP1E0XBN9T4B10E9QMR7XMFJPMA19D77WY3KP2QKC.self-listing-helper-v3a?chain=mainnet)

The `self-listing-helper-v3a` contract enables the creation of trading pools on the ALEX DEX through two distinct mechanisms: a permissioned flow and a permissionless flow. In both cases, pools are formed by pairing a governance-approved anchor token (`token-x`) with a listed token (`token-y`), which is the asset being introduced for trading.

This dual approach gives projects the flexibility to choose between a guided listing process or a fully autonomous, on-chain setup — all while using the same core infrastructure.

## Permissioned

> Pools can be created by following a guided process through the ALEX UI. The user supplies a `token-y` that must be pre-approved by governance before the pool can be initialized.

Reviewer: ```suggestion
Pools can be created by following a guided process through the [ALEX Lab UI](https://app.alexlab.co/self-service-listing). The user supplies a `token-y` that must be pre-approved by governance before the pool can be initialized.
```

## Permissionless

> Users can list a new `token-y` without requiring a prior approval step by deploying a wrapper contract that matches a governance-approved template. The contract includes a verification system that reconstructs and validates the original deployment transaction using Merkle proofs and block data. Once verification succeeds, the token is dynamically approved and the pool is created.

Reviewer: Acá agregaría un link en algún momento a la librería `clarity-stacks` también.

- `executor-dao`: calls are made to verify whether a certain contract-caller is designated as an extension.
- `liquidity-locker`: this contract is used to manage post-creation liquidity settings. It allows the contract to lock, burn, or later claim LP tokens based on the pool creator’s configuration.

> - `clarity-stacks`: this contract is used to verify that a wrapper contract was properly deployed on the Stacks blockchain. It checks that the contract was mined and included in a valid block. This system is built on top of Marvin Janssen’s `clarity-stacks` proof framework.

Reviewer: Agregar link al repositorio acá :)
